pkg/controllers/management/gke/gke_cluster_handler.go

```diff
@@ -0,0 +1,519 @@
+			}
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		}
+		logrus.Infof("waiting for cluster GKE [%s] pre-create failure to be resolved", cluster.Name)
+		return e.SetFalse(cluster, apimgmtv3.ClusterConditionProvisioned, failureMessage)
+	}
+}
+
+// setInitialUpstreamSpec builds a view of the upstream cluster and adds it to the status of the cluster resource
+func (e *gkeOperatorController) setInitialUpstreamSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	logrus.Infof("setting initial upstreamSpec on cluster [%s]", cluster.Name)
+	cluster = cluster.DeepCopy()
+	upstreamSpec, err := clusterupstreamrefresher.BuildGKEUpstreamSpec(e.SecretsCache, cluster)
+	if err != nil {
+		return cluster, err
+	}
+	cluster.Status.GKEStatus.UpstreamSpec = upstreamSpec
+	return e.ClusterClient.Update(cluster)
+}
+
+// updateGKEClusterConfig updates the GKEClusterConfig object's spec with the cluster's GKEConfig if they are not equal.
+func (e *gkeOperatorController) updateGKEClusterConfig(cluster *mgmtv3.Cluster, gkeClusterConfigDynamic *unstructured.Unstructured, spec map[string]interface{}) (*mgmtv3.Cluster, error) {
+	list, err := e.DynamicClient.Namespace(namespace.GlobalNamespace).List(context.TODO(), v1.ListOptions{})
+	if err != nil {
+		return cluster, err
+	}
+	selector := fields.OneTermEqualSelector("metadata.name", cluster.Name)
+	w, err := e.DynamicClient.Namespace(namespace.GlobalNamespace).Watch(context.TODO(), v1.ListOptions{ResourceVersion: list.GetResourceVersion(), FieldSelector: selector.String()})
+	if err != nil {
+		return cluster, err
+	}
+	gkeClusterConfigDynamic.Object["spec"] = spec
+	gkeClusterConfigDynamic, err = e.DynamicClient.Namespace(namespace.GlobalNamespace).Update(context.TODO(), gkeClusterConfigDynamic, v1.UpdateOptions{})
+	if err != nil {
+		return cluster, err
+	}
+
+	// GKE cluster and node pool statuses are not always immediately updated. This cause the GKEConfig to
+	// stay in "active" for a few seconds, causing the cluster to go back to "active".
+	timeout := time.NewTimer(10 * time.Second)
+	for {
+		select {
+		case event := <-w.ResultChan():
+			gkeClusterConfigDynamic = event.Object.(*unstructured.Unstructured)
+			status, _ := gkeClusterConfigDynamic.Object["status"].(map[string]interface{})
+			if status["phase"] == "active" {
+				continue
+			}
+
+			// this enqueue is necessary to ensure that the controller is reentered with the updating phase
+			e.ClusterEnqueueAfter(cluster.Name, enqueueTime)
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		case <-timeout.C:
+			cluster, err = e.recordAppliedSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+	}
+}
+
+// recordCAAndAPIEndpoint reads the GKEClusterConfig's secret once available. The CA cert and API endpoint are then copied to the cluster status.
+func (e *gkeOperatorController) recordCAAndAPIEndpoint(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
```

Comment:
It also looks like this function could be moved to `clusteroperator/utils.go`.

Reply:
done